rs/bitcoin/kyt/btc_kyt_canister.did

```diff
@@ -9,8 +9,33 @@ type CheckAddressArgs = record {
 
 type CheckAddressResponse = variant { Passed; Failed };
 
+type CheckTransactionArgs = record { txid : blob };
+
+// The result of a check_transaction call.
+type CheckTransactionResponse = variant {
```

Comment:
I think it would make sense to clearly distinguish the Ok case from any error case. What about having the Candid version of `Result<(), CheckTransactionError>` so that `check_transaction` has the following signature

```rust
async fn check_transaction(args: CheckTransactionArgs) -> Result<(), CheckTransactionError> 
```

Reply:
Thanks for the suggestions! I've refactored the code to introduce an error type with elaborated details. Please have a look again. Thanks!

Reply:
Thanks for the changes! I think having a `Result` as returned type definitely makes sense. One thing I would strive for is to have a single variant as `Ok` case (i.e., `()`) and everything else as errors that potentially need to be handled by the caller (whereas if the caller receives an `Ok(())` it knows that everything is fine and nothing needs to be handled). Maybe I'm overlooking something but I have the impression something like this would bring some clarity regarding which error can be retried. WDYT?

```rust
async fn check_transaction(
    args: CheckTransactionArgs,
) -> Result<(), CheckTransactionError> {
...
}

#[derive(CandidType, Debug, Deserialize, Serialize)]
pub enum CheckTransactionError {
    /// KYT failed. No need to retry, as the result is final.
    Failed(KytFailedStatus),
    /// Transaction status is not determined yet. Caller should call again.
    Pending(KytPendingStatus),
    /// Caller should call with a minimum of 40 billion cycles.
    NotEnoughCycles
}

#[derive(CandidType, Debug, Deserialize, Serialize)]
pub enum KytFailedStatus {
    // When check finishes and one or more input addresses failed KYT.
    Failed(Vec<Address>),
    /// Response size is too large (> `RETRY_MAX_RESPONSE_BYTES`) when fetching the transaction data of a txid.
    ResponseTooLarge { txid: Vec<u8> },
    /// Transaction is invalid: covers previous `Address`, `Tx`, `Txid` and `TxidMismatch` errors.
    InvalidTransaction(String)
}

#[derive(CandidType, Debug, Deserialize, Serialize)]
pub enum KytPendingStatus {
    /// More work to be done, and the caller should call again.
    Pending,
    /// The service is experiencing high load, and the caller should call again later.
    HighLoad,
    /// Transient internal error; retrying may cause the operation to succeed. Covers previous `Rejected` error
    TransientInternalError(String)
}
```

Reply:
> both Pass and Fail represent a final answer, which is something the rest cannot give.

yes makes sense.

> So I do not think Fail should be part of Err.

I think it depends a bit on the semantic of the function. We basically have 3 types of outcome (with possible sub-variations therein):

1. Final yes
2. Final no
3. Don't know

and we have currently 2 suggestions of grouping things: 

1. final yes/no together as `Ok`; Don't know as `Err`
2. final yes as `Ok`; Final no/ Don't know together as `Err`

I would say that the 2nd grouping is fairly common for methods named something like `check_` or `validate_` since they typically return a result of type `Result<(), SomeError>` and many callers that only care about ok result just call it as `check_.is_ok()...`).

If we want to go with the 1st grouping (which is also totally meaningful), maybe it would be better to call that method differently, e.g. `kyt_status`?

Reply:
> 1. Final yes
> 2. Final no
> 3. Don't know

Maybe it is wrong to try to fit 3 things into 2 slots? I do see that `Result` type can be convenient because of its member functions, but at its core it is just a enum of 2 branches. By not using the `Result` type, we won't even have this bike shedding. WDYT?

Also I would like to get opinion from @THLO, because he originally designed the interface, and I don't want to deviate too much from it.

Reply:
Discussed with @THLO and @ninegua :

1. Return types is an enum with 3 top-level variants (final yes, final no, unknown), no `Result`
2. Name of method to be defined later (`check_` is usually associated with a `Result` return type)